src/passes/OptimizeInstructions.cpp

```diff
@@ -1209,6 +1245,268 @@ struct OptimizeInstructions
+            }
+          } else if (left->op == Abstract::getBinary(type, Abstract::Add)) {
+            if (auto* constRight = left->right->dynCast<Const>()) {
+              //   (x + C) ^ -1   ==>   ~C - x
+              constRight->value =
+                constRight->value.xor_(Literal::makeFromInt32(-1, type));
+              left->op = Abstract::getBinary(type, Abstract::Sub);
+              std::swap(left->left, left->right);
+              return left;
+            } else if (auto* leftLeft = left->left->dynCast<Binary>()) {
+              if (leftLeft->left->type == left->right->type) {
+                auto* constLeftLeftRight = leftLeft->right->dynCast<Const>();
+                if (constLeftLeftRight &&
+                    constLeftLeftRight->value.getInteger() == -1LL) {
+                  //   ((x ^ -1) + y) ^ -1   ==>   x - y
+                  return Builder(*getModule())
+                    .makeBinary(Abstract::getBinary(type, Abstract::Sub),
+                                leftLeft->left,
+                                left->right);
+                }
+              }
+            } else if (auto* leftRight = left->right->dynCast<Binary>()) {
+              if (left->left->type == leftRight->left->type) {
+                //   (x + (y ^ -1)) ^ -1   ==>   y - x
+                if (auto* constLeftRightRight =
+                      leftRight->right->dynCast<Const>()) {
+                  if (constLeftRightRight->value.getInteger() == -1LL) {
+                    return Builder(*getModule())
+                      .makeBinary(Abstract::getBinary(type, Abstract::Sub),
+                                  leftRight->left,
+                                  left->left);
+                  }
+                }
+              }
+            }
+          } else if (left->op == Abstract::getBinary(type, Abstract::Xor)) {
+            //   (x ^ -1) ^ -1   ==>   x
+            if (auto* constLeftRigth = left->right->dynCast<Const>()) {
+              if (constLeftRigth->value.getInteger() == -1LL) {
+                return left->left;
+              }
+            }
+          }
+        }
+      }
+    } else if (auto* left = binary->left->dynCast<Binary>()) {
+      //  (x ^ -1) ^ (y ^ -1)   ==>   x ^ y
+      if (auto* right = binary->right->dynCast<Binary>()) {
+        if (left->left->type == right->left->type) {
+          if (auto* constLeftRight = left->right->dynCast<Const>()) {
+            if (constLeftRight->value.getInteger() == -1LL) {
+              if (auto* constRightRight = right->right->dynCast<Const>()) {
+                if (left->right->type == right->right->type) {
+                  if (constRightRight->value.getInteger() == -1LL) {
+                    binary->left = left->left;
+                    binary->right = right->left;
+                    return binary;
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    return nullptr;
+  }
```

Comment:
there is quite a lot here. I think it would really be better to break these PRs up into as small pieces as we can. The problem is that right now I can read part of it, but I have a bunch of comments already - which after you address, I will need to re-read the whole thing again.

instead, if each PR had only on specific optimization (I literally mean like one of the sub-cases here - something like 5-10 lines), and tests for just that, we could land pieces much more quickly, and reviewers would not need to read the same code again. It would also help bisection in the future.

I realize that will take more work, and I apologize for that, but I don't see another way to proceed here - there are so many of these PRs, and many are so large, it is just too overwhelming to review in big pieces.

Reply:
Another thing that might help is to have a clear order for the PRs. It's hard for me to read one, then another, and flip between them. If you decide a certain PR is the next one to review, and no more review of other ones until it lands, that would be excellent.

Reply:
Yes, sure. It sounds quite reasonable for me. So I split this PR into two parts and freeze my activity until all existing PRs aren't merged.